### ISSUE TITLE
Basic audit capabilities / Timestamps support

### ISSUE THREAD

**Issue by sestrella** [yesodweb/persistent#1071]
First, I just want to thank for such an amazing library. Before jumping into Haskell I was a Ruby developer for a quite a while, so you might have a sign where I'm trying to go with this proposal about adding auditable capabilities to persistent if you are familiar with ActiveRecord :) 

So here I go, I have been scratching my head about how to automatically handle timestamps with persistent and came up with the following idea for inserts and updates:

**Inserts**

Regarding inserts, I think partial application could be a quick solution, where auditable fields are handled by a type class:

```haskell
class Monad m => AuditableInsert partialRecord record m where
  auditableInsertRecord :: partialRecord -> m record

auditableInsert
  :: (AuditableInsert partialRecord record m, ...)
  => partialRecord
  -> ReaderT backend m (Key record)
auditableInsert partialRecord = lift (auditableInsertRecord partialRecord) >>= insert
```

A use case would look something similar to:

```haskell
type CreatedAt = UTCTime
type UpdatedAt = UTCTime

instance MonadIO m => AuditableInsert (CreatedAt -> UpdatedAt -> User) User where
  auditableInsertRecord = do
    createdAt <- liftIO getCurrentTime
    return $ f createdAt createdAt
```

In fact, a more generic instance could be declared alongside the type class:

```haskell
instance (PersistEntity record, MonadIO m)
  => AuditableInsert (CreatedAt -> UpdatedAt -> record) record where
```

**Updates**

For updates, type class would look pretty much the same, except it would return a list of `Update record`:

```haskell
class Monad m => AuditableUpdate record m where
   auditableUpdateFields :: m [Update record]

auditableUpdate
  :: (AuditableUpdate record m, ...)
  => Key record
  -> [Update record]
  -> ReaderT backend m ()
auditableUpdate key fields = do
  auditableFields <- lift auditableUpdateFields
  update key (fields ++ auditableFields) 
```

A use case would look something similar to:

```haskell
instance MonadIO m => AuditableUpdate Post m where
  auditableUpdateFields = do
    updatedAt <- liftIO getCurrentTime
    return [PostUpdatedAt =. updatedAt]
```

An extended use case would be adding some extra information, like who performed an update:

```haskell
instance MonadIO m => AuditableUpdate Post (ReaderT UserKey m) where
  auditableUpdateFields = do
    userId <- ask
    updatedAt <- liftIO getCurrentTime
    return [PostUpdatedBy =. userId, PostUpdatedAt =. updatedAt]
```

**Final thoughts**

The proposal above, still missing some edge cases like setting up auditable fields when passing the whole record to functions like `replace` or `repsert`, however, I just wanted to get some initial thoughts before going to deep into more technical details.

Even if this proposal gets rejected, I still think there is room to bundle it as a separate library, so I would appreciate any feedback, however, my initial thought was that auditable fields are a common feature for most application, on the other hand, taking into account all functions proposed are wrappers over the original persistent functions, it would make it easier to keep them in sync if both reside in the same code base.

**Comment by parsonsmatt** [comment]
I've actually come up with a proposal to fix this for `persistent` in a little more general manner: https://github.com/yesodweb/persistent/issues/1037#issuecomment-607562307

tl;dr: the `User` type exactly reflects the database schema, then we have `NewUser` which only has the required fields to insert, and `PartialUser` which allows you to specify the optional/database-defaulted values.

**Comment by sestrella** [comment]
@parsonsmatt - Thank for your rapid response, I reviewed that one before opening this issue, however, I decided to submit a proposal because I think the functionality we are discussing here is slightly different from default values, in the sense that once a record is stored in the DB, further operations like updates or even delete - soft delete - operations could include some extra auditable information. I think the new design proposed for persistent 3.0 - which looks like a huge improvement - removes the need of declaring a type class like `AuditableInsert`, but in order to keep auditable fields in sync for the rest of operations, it might be necessary to add some wrappers over existing functions. What do you think?

**Comment by parsonsmatt** [comment]
Oh, I see! So it's more about, "Here's a wrapper for `insert` / `update` that also performs some contex/record-dependent modification." Eg `updatedAt :: UTCTime` would be performed by the app layer, instead of a database trigger/default.

That's a cool idea! I've been wanting to write something similar that would let you use `persistent` with immutable tables (eg `update` is *actually* an `insert` to preserve the history). I think adding it as a library on top would be best.

**Comment by MaxGabriel** [comment]
I'm a big fan of doing this with SQL triggers, since otherwise raw SQL won't update timestamps. Should you want to go that approach, here's what we're using for it:

```
  CREATE OR REPLACE FUNCTION create_timestamps()   
    RETURNS TRIGGER AS $$
    BEGIN
      NEW.created_at = now();
      NEW.updated_at = now();
      RETURN NEW;   
    END;
    $$ language 'plpgsql';

  SQL

  execute <<-SQL
  CREATE OR REPLACE FUNCTION update_timestamps()   
    RETURNS TRIGGER AS $$
    BEGIN
      NEW.updated_at = now();
      RETURN NEW;   
    END;
    $$ language 'plpgsql';

  SQL

  execute <<-SQL
  CREATE TABLE users (
    id bigserial NOT NULL PRIMARY KEY,
    email text NOT NULL UNIQUE,
    password bytea NOT NULL,
    created_at timestamp with time zone NOT NULL,
    updated_at timestamp with time zone NOT NULL
  );

  CREATE TRIGGER users_insert BEFORE INSERT ON users FOR EACH ROW EXECUTE PROCEDURE create_timestamps();
  CREATE TRIGGER users_update BEFORE UPDATE ON users FOR EACH ROW EXECUTE PROCEDURE update_timestamps();
```

On the Haskell side, we either don't include the timestamp columns on our Persistent models, or if we do, we insert a `UTCTime (fromGregorian 1 1 1) 0` and let the database fill the values (hacky looking, admittedly, but works great)

**Comment by sestrella** [comment]
@MaxGabriel
